src/librustc_typeck/astconv.rs

```diff
@@ -415,58 +415,101 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 param_counts.lifetimes,
                 arg_counts.lifetimes,
                 0,
                 &mut unexpected_spans,
                 explicit_late_bound == ExplicitLateBound::Yes,
-            )
-            .and(arg_count_correct);
+            );
         }
+
         // FIXME(const_generics:defaults)
+        let mut const_count_correct = Ok(());
         if !infer_args || arg_counts.consts > param_counts.consts {
-            arg_count_correct = check_kind_count(
+            const_count_correct = check_kind_count(
                 "const",
                 param_counts.consts,
                 param_counts.consts,
                 arg_counts.consts,
                 arg_counts.lifetimes + arg_counts.types,
                 &mut unexpected_spans,
                 false,
-            )
-            .and(arg_count_correct);
+            );
         }
+
         // Note that type errors are currently be emitted *after* const errors.
+        let mut type_count_correct = Ok(());
         if !infer_args || arg_counts.types > param_counts.types - defaults.types - has_self as usize
         {
-            arg_count_correct = check_kind_count(
+            type_count_correct = check_kind_count(
                 "type",
                 param_counts.types - defaults.types - has_self as usize,
                 param_counts.types - has_self as usize,
                 arg_counts.types,
                 arg_counts.lifetimes,
                 &mut unexpected_spans,
                 false,
-            )
-            .and(arg_count_correct);
+            );
         }
 
+        // Emit a help message if it's possible that a type could be surrounded in braces
+        if let Err((c_mismatch, Some(ref mut _const_err))) = &mut const_count_correct {
+            if let Err((t_mismatch, Some(ref mut type_err))) = &mut type_count_correct {
+                if *t_mismatch < 0 && *c_mismatch <= -*t_mismatch {
+                    for i in 0..*c_mismatch as usize {
+                        let arg = &args.args[arg_counts.lifetimes + i];
+                        match arg {
+                            // Ignore anything other than paths
+                            GenericArg::Type(hir::Ty {
+                                kind: hir::TyKind::Path { .. }, ..
+                            }) => {}
+                            _ => continue,
+                        }
+                        let path = tcx.sess.source_map().span_to_snippet(arg.span());
+                        let path = if let Ok(path) = path {
+                            path
+                        } else {
+                            continue;
+                        };
+                        type_err.help(&format!(
+                            "If this generic argument was intended as a const parameter, \
+                            try surrounding it with braces: `{{ {} }}`",
+                            path,
+                        ));
```

Comment:
This is so nice already :sparkles:

Just realized that it might be even better to use a `multipart_suggestion` here, replacing `arg.span().shrink_to_lo()` with `{{ ` and `arg.span().shrink_to_hi()` with ` }}`